Feature update repository

Vasilijez opened the pull request:
Closes #27.

A regular user or admin can update the repository visibility or description. They can make an update only if they are either the repo owner or a member of a team that has admin permissions for the observed repo.

The functionalities are implemented based on Docker's approach.

Vasilijez committed:
feat: Add API to update repository description

Vasilijez committed:
feat: Add API for updating repository visibility

Vasilijez committed:
refactor: Use refactored function `update_repo_by_id` in `update_repo_desc_by_id` API

Vasilijez committed:
fix: Correct typos

Vasilijez committed:
feat: Add `can_update` attribute in `get_repo_by_canonical_name` API

Using this approach, we can elegantly check whether a user is allowed to edit a repository. This is necessary to dynamically show or hide editing elements.

Vasilijez committed:
refactor: Simplify commonly used permission checks for update-related endpoints

Vasilijez committed:
feat: Add UI elements for updating repo description and visibility

Enhance the `Overview` tab on the `RepoPage.tsx` to allow updating the repository description and implement the `Settings` tab for managing repository visibility.

Vasilijez committed:
feat: Finish `Visibility settings` card

Vasilijez committed:
feat: Finalize description editing

Vasilijez committed:
feat: Remove unnecessary check in 'user_has_update_permission' if user is already an organization owner

Vasilijez committed:
test: Add initial unit tests for 'user_has_update_permission'

Vasilijez committed:
test: Finish unit tests for `user_has_update_permission` and `update_repo_by_id`

Vasilijez committed:
test: Add initial integration test for `update_repo_by_id`

Vasilijez committed:
test: Improve integration test for `update_repo_by_id`

Vasilijez committed:
fix: Organization owner doesn't have admin editing permissions for all repositories by default

Vasilijez committed:
test: Finish integration test for `update_repo_by_id`

magley commented on server/app/api/repo/repo_service.py:
```diff
@@ -7,15 +7,57 @@
+            repo = self.repo_repo.set_visibility(repo, dto.public)
+        return repo
+
+    def _is_user_org_member_with_admin_permissions(self, user_id: int, repo: Repository):
+        user_is_organization_member = user_id in [member.user_id for member in repo.organization.members]
+        
+        if not user_is_organization_member:
+            return False
+
+        ''' Find all teams belonging to the organization '''
+        teams = self._get_all_teams_by_org(repo.organization_id)
+
+        for team in teams:
+            ''' Check if the team has admin permissions for the observed repository '''
+            for permission in team.permissions:
+                if permission.repo_id == repo.id and permission.kind == TeamPermissionKind.admin:
+                    ''' Check if the member belongs to the team '''
+                    if user_id in [member.user_id for member in team.members]:
+                        return True
```
Add a `return False` after the for loop in case none of the teams match the criteria.

Vasilijez replied:
Nice catch.

magley commented on client/src/components/RepoSettings.tsx:
```diff
@@ -0,0 +1,57 @@
+                        </span>
+                    </Card.Title>
+                    {error && <Alert variant="danger">{error}</Alert>}
+                    <p className="text-muted mt-3">
+                        {props.repo.public ? "This repository is public." : "This repository is private."}
+                    </p>
```
Could you bold the `public` and `private` text, and add the `bi-lock` icons here instead of the `<h5>` above? Also, make the icon size match the text size when you do.

magley commented on client/src/components/RepoSettings.tsx:
```diff
@@ -0,0 +1,57 @@
+                    <p className="text-muted mt-3">
+                        {props.repo.public ? "This repository is public." : "This repository is private."}
+                    </p>
+                    <Button variant="outline-primary" onClick={changeVisibility}>
+                        Change visibility
+                    </Button>
```
Instead of the generic `Change visibility`, put something more descriptive like "change to private", "make private" etc.

Vasilijez replied:
Haha, I added dynamic button text that changes from 'Make private' to 'Make public' and then removed it—it seemed like too much for me. I will add it back.

magley commented on client/src/components/RepoSettings.tsx:
```diff
@@ -0,0 +1,57 @@
+            props.repoStateChanger({
+                ...props.repo,
+                ...res.data,
+            });
+            addToast(`Updated repository ${props.repo.name}`, ToastType.success);
+            setError('');
```
Since this function _exclusively_ updates the visibility of the repository, make the toast message more descriptive as well ("{reponame} is now a private repository" or something like that).

magley commented on client/src/components/RepoOverview.tsx:
```diff
@@ -1,16 +1,66 @@
+            props.repoStateChanger({
+                ...props.repo,
+                ...res.data,
+            });
+            addToast(`Updated repository ${props.repo.name}`, ToastType.success);
+            setError('');
```
Since this function exclusively updates the description of the repository, make the toast message more descriptive as well ("Updated the description of {reponame}" or something like that).

magley commented on client/src/api/repo.api.ts:
```diff
@@ -1,5 +1,6 @@
 import { AxiosResponse } from "axios";
 import { axiosInstance } from "../util/http";
+import { dot } from "node:test/reporters";
```
This isn't necessary. I'm guessing you mistyped `dto` and your IDE imported `dot`.

magley commented on client/src/pages/RepoPage.tsx:
```diff
@@ -123,23 +124,34 @@ export const RepositoryPage: React.FC = () => {
                 <Tab.Content>
                     <Tab.Pane eventKey="overview">
-                        {repo && <RepoOverview isActive={key === 'overview'} repo={repo} />}
+                        {repo && <RepoOverview isActive={key === 'overview'} repo={repo} repoStateChanger={setRepo} />}
                     </Tab.Pane>
                     <Tab.Pane eventKey="tags">
                         <RepoTags isActive={key === 'tags'} />
                     </Tab.Pane>
+                    <Tab.Pane eventKey="settings">
+                        {repo && <RepoSettings isActive={key === 'settings'} repo={repo} repoStateChanger={setRepo} />}
+                    </Tab.Pane>
```
Is there a better way to name `repoStateChanger`? I'm not too familiar with React's state hooks so I don't know if this is the standard way to do it, but if not, consider just putting `setRepo`.

Vasilijez replied:
Honestly, I'm not sure. I didn’t follow any specific naming heuristics for React's state hook, `setRepo` seems fine to me.

magley reviewed (changes requested):
Seems okay, it's a simple enough feature anyway.
The web UI looks nice.
Most of my comments are regarding the web app.
I didn't dive into the tests as much, a lot of stuff will change when we integrate access control from develop into this branch.

Speaking of which, make sure to resolve all conflicts.

Vasilijez committed:
feat: Improve user interface

Vasilijez committed:
chore: Mistyped import

Vasilijez committed:
Merge branch 'develop' into feature-update-repository

Vasilijez committed:
feat: Replace old access control methodology with `access_control_service.py`

Vasilijez committed:
test: Update tests

Vasilijez committed:
fix: Add `spec` parameter to `MagicMock` constructors and update `team_permission` attribute

Tests for `update_repo_by_id` generate the following error:
`AttributeError: 'TeamPermission' object has no attribute 'permission'.`

Truly, `permission` attribute doesn't exist, but tests in `test_access_control.py` are successful. This is because `spec=Repository` is not used in the `MagicMock` constructor for the `repo` variable in `test_access_control.py`. Consequently, without this, any attribute can be defined.

I have updated all potentially problematic constructors to address this issue.

Vasilijez commented:
Don't forget to review [22b302e](https://github.com/magley/mocker-hub/pull/30/commits/22b302e0ed0115a0847352b26865f26ef0475047).

magley commented on server/app/api/access_control/access_control_service.py:
```diff
@@ -110,7 +110,7 @@ def has_write_access(self, user_id: int | None, repo_id: int) -> bool:
 
         for team_permission in team_permissions:
-            if team_permission.permission in [TeamPermissionKind.read_write, TeamPermissionKind.admin]:
+            if team_permission.kind in [TeamPermissionKind.read_write, TeamPermissionKind.admin]:
                 if self.team_repo.find_member(team_permission.team_id, user_id) is not None:
                     return True
```
Good catch, I guess I missed this because `add_dummy_data()` doesn't create teams or team permissions, so this for-loop had never executed when I was doing `docker push` manual testing.

magley commented on server/app/api/repo/repo_service.py:
```diff
@@ -18,6 +19,22 @@ def __init__(self, session: Session):
+    # use them because of circular dependencies (FastAPI does not support it).
+    
+    def _update_repo_attribute(self, repo: Repository, dto: RepositoryDescUpdateDTO | RepositoryVisibilityUpdateDTO) -> Repository:
+        if dto is None:
+            raise AccessDeniedException(f"Repository {repo.id} cannot be updated with a None value")
+        elif isinstance(dto, RepositoryDescUpdateDTO):
```
Instead of `AccessDeniedException`, how about something like BadRequest or InvalidPayload or something like that?

Vasilijez committed:
fix: Change exception type
